datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -186,6 +186,12 @@ fn signature(lhs: &DataType, op: &Operator, rhs: &DataType) -> Result<Signature>
             } else if let Some(numeric) = mathematics_numerical_coercion(lhs, rhs) {
                 // Numeric arithmetic, e.g. Int32 + Int32
                 Ok(Signature::uniform(numeric))
+            } else if let Some((new_lhs, new_rhs, ret)) = resolve_ints_to_intervals(lhs, rhs) {
```

Comment:
we're in `Plus | Minus | Multiply | Divide | Modulo => {` branch
but the new logic should apply 

- to `+` symmetrically
- to `date - int`, but likely not to `int - date` (at least per PostgreSQL)
- not at all to other operators

so what about passing the operator to the new function and doing the work for `+` and `-` (so that code already identifies asymmetric treatment)